### ISSUE TITLE
prepublishOnly running after publish.

### ISSUE THREAD

**Issue by MicahZoltu** [npm/npm#17305]
Re-opening https://github.com/npm/npm/issues/15454 per bot recommendation.

I'm opening this issue because:

 npm is crashing.
[x ] npm is producing an incorrect install.
 npm is doing something I don't understand.
 Other (see below for feature requests):
What's going wrong?

I have a prepublishOnly script that takes my source and transpiles, lints and runs tests before publishing. However I frequently find that I will run npm publish which will run the prepublish script, however when I pull the package down the dist/ directory has not been changed. Currently, I've been working around this by manually running npm run prepublishOnly npm publish.

How can the CLI team reproduce the problem?

Write an prepublishOnly script that does some non-trivial compilation that takes a few seconds. Run npm publish and check the source to see if the script actually finished before the package was published to the registry. It's not super consistent either: sometimes it publishes right away, sometimes it runs the clean script I have set up (which deletes the dist directory entirely) and then publishes without dist; other times it seems to work as intended.

I think this issue on travis may be related to the issue I'm having and the user misattributed the error to travis: travis-ci/travis-ci#6953
He also has a repro there but I have not tried it.
Otherwise, you can try my package as a repro: https://github.com/tbtimes/lede/tree/update

supporting information:

npm -v prints: 4.0.3
node -v prints: v6.9.2
npm config get registry prints: https://registry.npmjs.org/
Windows, OS X/macOS, or Linux?: OSX

**Comment by caballus** [comment]
same on Ubuntu 16.04
npm -v: 4.6.1
node -v: v7.8.0

**Comment by Jmeyering** [comment]
Yes the `prepublishOnly` script currently runs AFTER the pack step, and according to this comment

https://github.com/npm/npm/issues/15147#issuecomment-288599684

Could be considered a bug. I am also running into this, where I want to run a transpile step before I publish but prepublishOnly does not allow for that. I have to manually build before I run `npm publish`

**Comment by whymarrh** [comment]
This is an issue with npm@4. npm@5 fixed this. From [the v5.0.0 changelog](https://github.com/npm/npm/blob/0cc9d89ed2d46745f91d746fda9d205fd39d3daa/CHANGELOG.md#v500-2017-05-25):

> - `prepublishOnly` now runs before the tarball to publish is created, after `prepare` has run.
